File: src/autosharing/output.py

from model import Solution
import csv

class ProcessOutput:
    def __init__(self, solution: Solution) -> None:
        self.car_to_zone =  solution.car_to_zone
        self.req_to_car = solution.req_to_car
        self.cost = solution.cost
        self.writeCSV()

    def writeCSV(self):
        with open('solution.csv', 'w', newline='') as csv_file:
            writer = csv.writer(csv_file)
            writer.writerow(self.cost)
            writer.writerow("+Vehicle assignments")
            for i in range(len(self.car_to_zone)):
                car = "car" + str(i)
                zone = "z" + str(self.car_to_zone[i])
                writer.writerow(car + ';' + zone)

            writer.writerow("+Assigned requests")
            for j in range(len(self.req_to_car)):
                if(self.req_to_car[j] != -1):
                    req = "req" + str(j)
                    car = "car" + str(self.req_to_car[j])
                    writer.writerow(req + ';' + car)

            writer.writerow("+Unassigned requests")
            for k in range(len(self.req_to_car)):
                if(self.req_to_car[k] == -1):
                    req = "req" + str(k)
                    writer.writerow(req)

File: src/output.rs
use crate::model::SolutionModel;
use std::fs;

pub fn ouput_solution(output_file: String,
                  solution: SolutionModel) -> Result<(), String> {
    let mut output = String::new();
    let cost = solution.cost;
    output.push_str(&format!("{cost}\n"));
    output.push_str("+Vehicle assignments\n");
    for i in 0..solution.car_to_zone.len() {
        let zone = solution.car_to_zone[i];
        output.push_str(&format!("car{i};z{zone}\n"));
    }
    output.push_str("+Assigned requests\n");
    for i in 0..solution.req_to_car.len() {
        if solution.req_to_car[i] != -1 {
            let car = solution.req_to_car[i];
            output.push_str(&format!("req{i};car{car}\n"));
        }
    }
    output.push_str("+Unassigned requests\n");
    for i in 0..solution.req_to_car.len() {
        if solution.req_to_car[i] == -1 {
            output.push_str(&format!("req{i}\n"));
        }
    }
    fs::write(output_file, output)
        .map_err(|x| format!("io error writing solution: {x}"))?;
    return Ok(());
}
